rwengine/src/audio/SoundManager.cpp

```diff
@@ -173,6 +178,226 @@ void SoundManager::SoundSource::loadFromFile(const std::string& filename) {
     avformat_close_input(&formatContext);
 }
 
+/// Structure for input data
+struct InputData {
+    uint8_t *ptr = nullptr;
+    size_t size{}; ///< size left in the buffer
+};
+
+/// Low level function for copying data from handler (opaque)
+/// to buffer.
+static int read_packet(void *opaque, uint8_t *buf, int buf_size) {
+    auto* input = reinterpret_cast<InputData*>(opaque);
+    buf_size = FFMIN(buf_size, input->size);
+    /* copy internal data to buf */
+    memcpy(buf, input->ptr, buf_size);
+    input->ptr  += buf_size;
+    input->size -= buf_size;
+    return buf_size;
+}
+
+void SoundManager::SoundSource::loadSfx(const rwfs::path& path, const size_t& index, const bool asWave) {
+    // Allocate audio frame
+    AVFrame* frame = av_frame_alloc();
+    if (!frame) {
+        RW_ERROR("Error allocating the audio frame");
+        return;
+    }
+
+    /// Now we need to prepare "custom" format context
+    /// We need sdt loader for that purpose
+    LoaderSDT sdt{};
+    sdt.load(path / "audio/sfx");
+
+    std::unique_ptr<char[]> raw_sound = sdt.loadToMemory(index, asWave);
+    if (!raw_sound) {
+        av_frame_free(&frame);
+        RW_ERROR("Error loading sound");
+        return;
+    }
+
+    /// Prepare input
+    InputData input{};
+    input.size = sizeof(WaveHeader) + sdt.assetInfo.size;
+    auto inputDataStart = std::make_unique<uint8_t[]>(input.size); /// Store start ptr of data to be able freed memory later
+    input.ptr = inputDataStart.get();
+
+    /// Alocate memory for buffer
+    /// Memory freeded at the end
+    static constexpr size_t ioBufferSize = 4096;
+    auto  ioBuffer = static_cast<uint8_t*>(av_malloc(ioBufferSize));
+
+    /// Convert memory, in order to match required layout for ffmpeg
+    memcpy(input.ptr, raw_sound.get(), input.size);
```

Comment:
Why does this need to be copied? Can't the memory for raw_sound be given to the input?

Reply:
I don't like pointer casting, but memory layout should match. I'll check.